### DISCUSSION TITLE
osg::PositionAttitudeTransform equivalent class in vsg

### DISCUSSION THREAD

**Discussion by CarlSandstrom** [vsg-dev/VulkanSceneGraph#704] [Q&A] [answered]
I am currently rewriting code that uses OpenSceneGraph to use VulkanSceneGraph. Now I have some code that uses a class called _osg::PositionAttitudeTransform_ and it seems that it is a class that makes translation and rotation of subgraphs easy. However, I cannot find an equivalent class in VulkanSceneGraph. Is there a way I can achieve the same behavior in VulkanSceneGraph?

**Comment by robertosfield** [comment, marked as answer]
I haven't implemented an directly replacement for osg::PositionAttitudeTransform in the VSG, but I've implemented variations on this several times for different client projects by subclassing from [vsg::Transform](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/nodes/Transform.h) and override the transform method:

     virtual dmat4 transform(const dmat4& mv) const = 0;

The are currently two subclasses from Transform in the core VSG that could be used as a guide to what is required:

1. vsg::MatrixTransform [jheader] (https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/nodes/MatrixTransform.h) & [source](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/nodes/MatrixTransform.cpp)
2. vsg::AbsoluteTransform [header](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/include/vsg/nodes/AbsoluteTransform.h) & [source](https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/nodes/AbsoluteTransform.cpp)

Potentially we could include a vsg::PositionAttitudeTransform if it's something that will be widely used, or we could just leave this up to the user to create the transform subclasses that suit them best - for instance not everyone will need a scale and pivot point etc.

If you want to use existing VSG without subclassesing from vsg::Transform the way to do it would be to use a MatrixTransfom and set the matrix member directly i.e.

```
auto mt = vsg::MatrixTransform::create();
mt->matrix = vsg::translate(position) * vsg::rotate(attitude) * vsg::scale(s) * vsg::translate(-pivot);
```
This will be just fine if you just set the Transform node and then don't modify it, but if you need to keep updating the position and attitude then you'd end up forced to hold this members somewhere and keep applying them to the MatrixTransform when they are updated. In this case it might be easier to just implement a subclass of Transform,

**Reply by CarlSandstrom** [reply]
Thanks! I went with the subclassing and it turned out fine.
